File: modules/anon.py
import discord
from discord.ext import commands
import random
import re  # For URL detection
import asyncio  # For adding delay

class Anoncord(commands.Cog):
    def __init__(self, bot):
        self.bot = bot
        # Replace with a list of channel IDs for target channels
        self.anon_channel_ids = [
            1314503562960834571,  # Channel ID 1
            1314517087414124545,  # Channel ID 2
            # Add more channel IDs here
        ]
        self.delete_delay = 0.1  # Delay in seconds before deleting the original message

    def is_valid_url(self, url):
        # Basic URL validation
        regex = r"(http|https)://[^\s]+"
        return re.match(regex, url)

    @commands.command(name='anon', help='Send an anonymous message, file, or link to specified channels.')
    async def anonymous_message(self, ctx, *, message: str = None):
        # Add a slight delay before deleting the original message
        try:
            await asyncio.sleep(self.delete_delay)
            await ctx.message.delete()
        except discord.Forbidden:
            await ctx.send("I don't have permission to delete messages here, but sent anyway.")

        # Generate a random identifier for the anonymous message
        random_id = random.randint(1, 9999)

        # Prepare embed and files for the message
        embed = discord.Embed(description=f"### 🔏 [ANON-{random_id}] 🔏")
        file_attachments = []

        # Handle attachments (files/images)
        if ctx.message.attachments:
            for attachment in ctx.message.attachments:
                # Handle images
                if attachment.content_type and attachment.content_type.startswith("image"):
                    embed.add_field(name="📬", value=f"({attachment.url})", inline=False)
                    embed.set_image(url=attachment.url)  # Add first image to the embed preview
                else:
                    # Handle non-image files
                    file_attachments.append(await attachment.to_file())
                    embed.add_field(name="📦", value=f"[{attachment.filename}]({attachment.url})", inline=False)

        # Handle the message content
        if message:
            if self.is_valid_url(message):
                embed.add_field(name="📷", value=message, inline=False)
            else:
                embed.add_field(name="💬", value=message, inline=False)

        # Send the consolidated message to all specified channels
        if embed.fields or file_attachments:
            for channel_id in self.anon_channel_ids:
                anon_channel = self.bot.get_channel(channel_id)
                if anon_channel is None:
                    continue  # Skip if the channel is not found
                await anon_channel.send(embed=embed, files=file_attachments)
        else:
            await ctx.send("You must provide a message or attach a file/image to send.")

    @commands.command(name='anonfile', help='Alias for sending anonymous messages with files.')
    async def anonymous_file(self, ctx):
        await self.anonymous_message(ctx)

# Setup function to add the cog to the bot
async def setup(bot):
    await bot.add_cog(Anoncord(bot))

File: modules/anon.js
const { SlashCommandBuilder, EmbedBuilder, MessageFlags } = require('discord.js');
const { customId, showModal } = require('../lib/interactions');

const MODULE = 'anon';
const ANON_CHANNEL_IDS = [
  '1314503562960834571', // Channel ID 1
  '1314517087414124545'  // Channel ID 2
];

function isValidUrl(url) {
  try {
    new URL(url);
    return true;
  } catch {
    return false;
  }
}

// Assumes at least one of message/file is truthy — caller guarantees that.
async function sendAnon(client, message, file) {
  const randomId = Math.floor(Math.random() * 9999) + 1;
  const embed = new EmbedBuilder().setDescription(`### 🔏 [ANON-${randomId}] 🔏`);
  const files = [];

  if (file) {
    if (file.contentType && file.contentType.startsWith('image')) {
      embed.addFields({ name: '📬', value: `(${file.url})`, inline: false });
      embed.setImage(file.url);
    } else {
      files.push({ attachment: file.url, name: file.name });
      embed.addFields({ name: '📦', value: `[${file.name}](${file.url})`, inline: false });
    }
  }

  if (message) {
    embed.addFields({ name: isValidUrl(message) ? '📷' : '💬', value: message, inline: false });
  }

  for (const channelId of ANON_CHANNEL_IDS) {
    const channel = client.channels.cache.get(channelId);
    if (channel) await channel.send({ embeds: [embed], files });
  }
}

module.exports = {
  data: new SlashCommandBuilder()
    .setName('anon')
    .setDescription('Send an anonymous message, file, or link to specified channels.')
    .addStringOption(option =>
      option.setName('message')
        .setDescription('Message or URL to send')
        .setRequired(false)
    )
    .addAttachmentOption(option =>
      option.setName('file')
        .setDescription('File or image to send')
        .setRequired(false)
    ),
  async execute(interaction) {
    const message = interaction.options.getString('message');
    const file = interaction.options.getAttachment('file');

    if (message || file) {
      await sendAnon(interaction.client, message, file);
      await interaction.reply({ content: 'Sent anonymously!', flags: MessageFlags.Ephemeral });
      return;
    }

    // Neither option given: compose the text in a modal instead of erroring
    // out. showModal() must be the FIRST response to this interaction —
    // execute() hasn't replied yet at this point, so this is valid.
    const modalSubmit = await showModal(interaction, {
      id: customId(MODULE, 'compose'),
      title: 'Anonimowa wiadomość',
      fields: [{ id: 'message', label: 'Treść (tekst lub URL)', style: 'paragraph', required: true, maxLength: 2000 }]
    });
    if (!modalSubmit) return; // timed out

    const text = modalSubmit.fields.getTextInputValue('message');
    await sendAnon(interaction.client, text, null);
    await modalSubmit.reply({ content: 'Sent anonymously!', flags: MessageFlags.Ephemeral });
  }
};
